fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/StandardIndexMaintainer.java

```diff
@@ -345,18 +350,74 @@ public boolean skipUpdateForUnchangedKeys() {
                 state.store.countKeyValue(FDBStoreTimer.Counts.SAVE_INDEX_KEY, FDBStoreTimer.Counts.SAVE_INDEX_KEY_BYTES, FDBStoreTimer.Counts.SAVE_INDEX_VALUE_BYTES,
                         keyBytes, valueBytes);
             }
+            return AsyncUtil.DONE;
         }
     }
 
-    @Override
-    @Nonnull
-    public void updateUniquenessViolations(@Nonnull Tuple valueKey, @Nonnull Tuple primaryKey, @Nullable Tuple existingKey, boolean remove) {
+    protected <M extends Message> void checkUniqueness(@Nonnull FDBIndexableRecord<M> savedRecord, @Nonnull IndexEntry indexEntry) {
+        Tuple valueKey = indexEntry.getKey();
+        AsyncIterable<KeyValue> kvs = state.transaction.getRange(state.indexSubspace.range(valueKey));
+        Tuple primaryKey = savedRecord.getPrimaryKey();
+        final CompletableFuture<Void> checker = state.store.getContext().instrument(FDBStoreTimer.Events.CHECK_INDEX_UNIQUENESS,
+                AsyncUtil.forEach(kvs, kv -> {
+                    Tuple existingEntry = unpackKey(getIndexSubspace(), kv);
+                    Tuple existingKey = state.index.getEntryPrimaryKey(existingEntry);
+                    if (!TupleHelpers.equals(primaryKey, existingKey)) {
+                        if (state.store.isIndexWriteOnly(state.index)) {
+                            addUniquenessViolations(valueKey, primaryKey, existingKey);
+                            addUniquenessViolations(valueKey, existingKey, primaryKey);
+                        } else {
+                            throw new RecordIndexUniquenessViolation(state.index, indexEntry, primaryKey, existingKey);
+                        }
+                    }
+                }, getExecutor()));
+        state.store.getRecordContext().addCommitCheck(checker);
```

Comment:
Could we just return the future considering `updateOneKey` is now async?

Reply:
I believe the purpose of making it a pre-commit check is that it means that if someone violates a uniqueness constraint, then we don't accidentally commit the transaction.

I could see an argument being made, though, that we'd want to fail fast...if we can. I think we definitely want preserve the behavior where committing is not allowed if there's a uniqueness constraint violation caused by a record insert, though.

Reply:
(It's possibly worth memorializing this with a comment--something like, "add a pre-commit check to prevent accidentally committing and getting into an invalid state" or something.)

Reply:
Added the comment.